styx-common/src/main/java/com/spotify/styx/util/ReplayEvents.java

```diff
@@ -35,47 +36,80 @@ private ReplayEvents() {
     throw new UnsupportedOperationException();
   }
 
-  // TODO: fix NPath complexity
-  public static Optional<RunState> getBackfillRunState(
+  public static Optional<RunStateData> getBackfillRunStateData(
       WorkflowInstance workflowInstance,
       Storage storage,
       String backfillId) {
-    final SettableTime time = new SettableTime();
-    boolean backfillFound = false;
-
-    final SortedSet<SequenceEvent> sequenceEvents;
-    try {
-      sequenceEvents = storage.readEvents(workflowInstance);
-    } catch (IOException e) {
-      throw new RuntimeException(e);
-    }
 
+    final SortedSet<SequenceEvent> sequenceEvents = getSequenceEvents(workflowInstance, storage);
     if (sequenceEvents.isEmpty()) {
       return Optional.empty();
     }
 
+    final SettableTime time = new SettableTime();
     RunState restoredState = RunState.fresh(workflowInstance, time);
 
+    boolean backfillFound = false;
+
+    final SettableTime initialTime = new SettableTime();
+    final SettableTime latestTime = new SettableTime();
+
     // events are written after the datastore transition transaction is successfully
     // committed, so we can trust the sequence of events faithfully reflect the state
     // transition if they have all been successfully written to bigtable
     for (SequenceEvent sequenceEvent : sequenceEvents) {
       time.set(Instant.ofEpochMilli(sequenceEvent.timestamp()));
-      if ("triggerExecution".equals(EventUtil.name(sequenceEvent.event()))) {
+
+      final boolean triggerExecutionEventMet =
+          "triggerExecution".equals(EventUtil.name(sequenceEvent.event()));
+
+      if (triggerExecutionEventMet) {
         if (backfillFound) {
-          return Optional.of(restoredState);
+          break;
         }
+
         restoredState = RunState.fresh(workflowInstance, time);
+        initialTime.set(time.get());
```

Comment:
This will include all the time spent waiting for dependencies, which for scheduled executions of most workflows is a long time.

We should perhaps only measure the time spent running the container.